File: python_algorithm/boj18870.py
import sys
from io import StringIO


def binary_search(n, arr):
    start = 0
    end = len(arr) - 1
    mid_ptr = (start + end) // 2

    # start mid를 고려한다.
    # break point
    while start <= mid_ptr:

        if arr[mid_ptr] == n:
            return 1, start, mid_ptr, end
        elif arr[mid_ptr] > n:
            end = mid_ptr - 1
        else:
            start = mid_ptr + 1

        mid_ptr = (start + end) // 2

    return 0


def lower_bound(target, arr):
    start = 0
    end = len(arr)
    mid = (start + end) // 2

    while start < end:

        if arr[mid] >= target:
            end = mid
        else:
            start = mid + 1
        mid = (start + end) // 2

    return start

def upper_bound(target, arr):
    start = 0
    end = len(arr)
    mid = (start + end) // 2

    while start < end:

        if arr[mid] > target:
            end = mid
        else:
            start = mid + 1
        mid = (start + end) // 2

    return start
    #0  1  2 / target= 3
    # [0, 1]


    # s: 0 / mid: 4 / e: 8
    # s: 0 / mid: 2 / e: 4
    # s: 3 / mid: 3 / e: 4


# v[i - 1] < target <= v[i]
# v[i - 1] <= target < v[i]


def main() -> None:
    # N = int(sys.stdin.readline().rstrip())
    # arr = sorted(list(map(lambda x: int(x), sys.stdin.readline().rstrip().split(' '))))
    #      0  1  2  3  4  5  6  7  8
    arr = [1, 1, 1, 2, 2, 2, 2, 3, 4]

    res = upper_bound(3, arr)

    print(res)
    io = StringIO()

    # for n in target_nums:
    #     io.write(f"{binary_search(n, arr)}\n")
    #
    # print(io.getvalue())


if __name__ == '__main__':
    main()

File: rust_algorithm/examples/boj18870.rs
#![allow(unused)]
#![allow(non_snake_case)]

use std::io;
use std::io::Read;
use std::ptr::{hash, NonNull};

fn custom_read_line<'a>() -> String {
    let mut buf = String::new();

    // expect internal code
    /*
        pub fn expect(self, &str: message) -> {
            match self {
                Ok(t) => t,
                Err(e:E) => unwrap_failed(message, &e),
            }

        }
     */
    io::stdin().read_line(&mut buf).expect("When reading a data from buf");

    buf.trim_end().to_string()
}

fn lower_bound(arr: &Vec<i32>, target: i32) -> usize {
    let mut start = 0;
    let mut end = arr.len();
    let mut mid = (start + end) / 2;

    // let mut result = None;
    while start < end {
        if arr[mid] >= target {
            end = mid
        } else { // arr[mid] < target
            start = mid + 1
        }
        mid = (start + end) / 2;
    }

    start
}

fn upper_bound(arr: &Vec<i32>, target: i32) -> usize {
    let mut start: usize = 0;
    let mut end: usize = arr.len();
    let mut mid: usize = (start + end) / 2;

    // let mut result = None;
    while start < end {
        if arr[mid] <= target {
            start = mid + 1
        } else {
            end = mid
        }
        mid = (start + end) / 2;
    }

    start
}


fn binary_search(arr: &Vec<i32>, target: i32) -> u8 {
    let mut start: usize = 0;
    let mut end: usize = arr.len();
    let mut mid: usize = (start + end) / 2;

    let mut answer: u8 = 0;

    while start < end {
        if arr[mid] == target {
            answer = 1;
            break;
        } else if arr[mid] > target {
            end = mid;
        } else {
            start = mid + 1;
        }

        mid = (start + end) / 2;
    }

    answer
}

fn main() {
    let buf = custom_read_line();
    let N = buf.parse::<u32>().unwrap();

    let buf = custom_read_line();
    let s: Vec<&str> = buf.split(' ').collect();

    let mut itr = s.iter().map(|x| x.parse::<i32>().unwrap());
    let mut arr: Vec<_> = itr.collect();

    let target_vec: Vec<i32> = arr.clone();
    arr.sort();

    let mut i: usize = 0;
    let mut v: Vec<i32> = vec![];
    v.push(arr[i]);
    let mut upper_bound_idx = upper_bound(&arr, arr[i]);

    while upper_bound_idx < arr.len() {
        v.push(arr[upper_bound_idx]);

        upper_bound_idx = upper_bound(&arr, arr[upper_bound_idx]);
    }

    let mut answer = String::new();

    for target in target_vec {

        let idx = lower_bound(&v, target);
        answer.push_str(&format!("{} ", idx));
    }
    println!("{}", answer);
}
